modules/channel-messenger/src/backend/messenger.ts

```diff
@@ -0,0 +1,253 @@
+
+    this.bp.events.registerMiddleware({
+      description: 'Sends outgoing messages for the messenger channel',
+      direction: 'outgoing',
+      handler: this._handleOutgoingEvent.bind(this),
+      name: 'messenger.sendMessages',
+      order: 200
+    })
+  }
+
+  getMessengerClient(botId: string): MessengerClient {
+    if (this.messengerClients[botId]) {
+      return this.messengerClients[botId]
+    }
+
+    this.messengerClients[botId] = new MessengerClient(botId, this.bp, this.http)
+    return this.messengerClients[botId]
+  }
+
+  // See: https://developers.facebook.com/docs/messenger-platform/webhook#security
+  private async _verifySignature(req, res, buffer) {
+    const client = this.getMessengerClient(req.params.botId)
+    const config = await client.getConfig()
+    const signatureError = new Error("Couldn't validate the request signature.")
+
+    if (!/^\/webhook/i.test(req.path)) {
+      return
+    }
+
+    const signature = req.headers['x-hub-signature']
+    if (!signature) {
+      throw signatureError
+    } else {
+      const [, hash] = signature.split('=')
+      const expectedHash = crypto
+        .createHmac('sha1', config.verifyToken)
+        .update(buffer)
+        .digest('hex')
+
+      if (hash != expectedHash) {
+        throw signatureError
+      }
+    }
+  }
+
+  private async _handleIncomingMessage(req, res) {
+    const body = req.body
+    const botId = req.params.botId
+    const client = this.getMessengerClient(botId)
+
+    if (body.object !== 'page') {
+      res.sendStatus(404)
+      return
+    }
+
+    for (const entry of body.entry) {
+      // Will only ever contain one message, so we get index 0
+      const webhookEvent = entry.messaging[0]
+      const senderId = webhookEvent.sender.id
+
+      await client.sendAction(senderId, 'mark_seen')
+
+      if (webhookEvent.message) {
+        await this._sendEvent(botId, senderId, webhookEvent.message, { type: 'message' })
+      } else if (webhookEvent.postback) {
+        await this._sendEvent(botId, senderId, { text: webhookEvent.postback.payload }, { type: 'callback' })
+      }
+    }
+
+    res.status(200).send('EVENT_RECEIVED')
+  }
+
+  private async _sendEvent(botId: string, senderId: string, message, args: { type: string }) {
+    this.bp.events.sendEvent(
+      this.bp.IO.Event({
+        botId,
+        channel: 'messenger',
+        direction: 'incoming',
+        payload: message,
+        preview: message.text,
+        target: senderId,
+        ...args
+      })
+    )
+  }
+
+  private async _setupWebhook(req, res) {
+    const mode = req.query['hub.mode']
+    const token = req.query['hub.verify_token']
+    const challenge = req.query['hub.challenge']
+    const botId = req.params.botId
+
+    const client = this.getMessengerClient(botId)
+    const config = await client.getConfig()
+
+    if (mode && token && mode === 'subscribe' && token === config.verifyToken) {
```

Comment:
is there another mode than subscribe ? do we want to support it ?

Reply:
I assume there are other modes. I'm not sure its relevant to support other modes for features we don't currently use?